src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs

```diff
@@ -112,24 +114,60 @@ public override ValueTask<T> ReadAsync(CancellationToken cancellationToken)
 
             public override bool TryRead([MaybeNullWhen(false)] out T item)
             {
-                UnboundedChannel<T> parent = _parent;
+                UnboundedChannel<T, TQueue> parent = _parent;
+                return parent._items.IsThreadSafe ?
+                    IsThreadSafe(parent, out item) :
+                    NotIsThreadSafe(parent, out item);
 
-                // Dequeue an item if we can
-                if (parent._items.TryDequeue(out item))
+                static bool IsThreadSafe(UnboundedChannel<T, TQueue> parent, [MaybeNullWhen(false)] out T item)
                 {
-                    CompleteIfDone(parent);
-                    return true;
+                    if (parent._items.TryDequeue(out item))
+                    {
+                        CompleteIfDone(parent);
+                        return true;
+                    }
+
+                    item = default;
+                    return false;
                 }
 
-                item = default;
-                return false;
+                static bool NotIsThreadSafe(UnboundedChannel<T, TQueue> parent, [MaybeNullWhen(false)] out T item)
+                {
+                    lock (parent.SyncObj)
+                    {
+                        if (parent._items.TryDequeue(out item))
+                        {
+                            CompleteIfDone(parent);
+                            return true;
+                        }
+                    }
+
+                    item = default;
+                    return false;
+                }
             }
 
-            public override bool TryPeek([MaybeNullWhen(false)] out T item) =>
-                _parent._items.TryPeek(out item);
+            public override bool TryPeek([MaybeNullWhen(false)] out T item)
+            {
+                UnboundedChannel<T, TQueue> parent = _parent;
+                return parent._items.IsThreadSafe ?
+                    parent._items.TryPeek(out item) :
+                    NotIsThreadSafe(parent, out item);
```

Comment:
This also fixes a bug in my previous PR that added the prioritized channel: TryPeek wasn't appropriately locking.

Reply:
Thanks for adding the comment! FWIW you have marked it as resolved and I almost missed it:

![image](https://github.com/dotnet/runtime/assets/6011991/35d9dcc2-c6c1-450a-af07-77f3ff4afe26)

Reply:
Sorry :)